src/BTVNanoCommissioning/workflows/BTA_producer.py

```diff
@@ -469,6 +518,250 @@ def process_shift(self, events, shift_name):
+                zeros + idx,
+                trkj_jetbased["category"],
+            )
+
+        # calculate track probability, based on IPsig and category
+        jpc = JPCalibHandler(self._campaign, isRealData, dataset)
+        trkj_jetbased["proba"] = jpc.calc_track_proba(
+            trkj_jetbased.btagSip3dSig,
+            ak.where(trkj_jetbased.category >= 0, trkj_jetbased.category, 0),
+        )
+
+        # then calculate JP/JBP
+
+        # (a) obtain track probabilities used to calculate JP
+        # only select positive/negative IP tracks for positive/negative JP tagger
+        # also requiring dR < 0.3
+        # reference code: https://github.com/cms-sw/cmssw/blob/CMSSW_13_0_X/RecoBTag/ImpactParameter/interface/TemplatedJetProbabilityComputer.h
+        trk_proba_pos = abs(trkj_jetbased["proba"])[
+            (trkj_jetbased.btagSip3dSig > 0)
+            & (trkj_jetbased.dr_jet < 0.3)
+            & ((trkj_jetbased.category >= 0) & (trkj_jetbased.category < 10))
+        ]
+        trk_proba_neg = abs(trkj_jetbased["proba"])[
+            (trkj_jetbased.btagSip3dSig < 0)
+            & (trkj_jetbased.dr_jet < 0.3)
+            & ((trkj_jetbased.category >= 0) & (trkj_jetbased.category < 10))
+        ]
+        Jet["Proba"] = -np.log10(jpc.calc_jet_proba(trk_proba_pos)) / 4.0
+        Jet["ProbaN"] = -np.log10(jpc.calc_jet_proba(trk_proba_neg)) / 4.0
+
+        # (b) obtain track probabilities used to calculate positive JBP tagger
+        # use two sets: probabilities and probabilitiesB (select 4 highest IPsig tracks)
+        # reference code: https://github.com/cms-sw/cmssw/blob/CMSSW_13_0_X/RecoBTag/ImpactParameter/interface/TemplatedJetBProbabilityComputer.h
+        trk_proba_all = abs(trkj_jetbased["proba"])[
+            ((trkj_jetbased.category >= 0) & (trkj_jetbased.category < 10))
+        ]
+        trk_proba_pos = abs(trkj_jetbased["proba"])[
+            (trkj_jetbased.btagSip3dSig > 0)
+            & ((trkj_jetbased.category >= 0) & (trkj_jetbased.category < 10))
+        ]
+        trk_proba_neg = abs(trkj_jetbased["proba"])[
+            (trkj_jetbased.btagSip3dSig < 0)
+            & ((trkj_jetbased.category >= 0) & (trkj_jetbased.category < 10))
+        ]
+        trk_Bproba_pos = ak.sort(trk_proba_pos)[
+            ak.local_index(trk_proba_pos) < 4
+        ]  # select up to 4 tracks
+        trk_Bproba_neg = ak.sort(trk_proba_neg)[
+            ak.local_index(trk_proba_neg) < 4
+        ]  # select up to 4 tracks
+
+        # for positive JBP tagger: probabilities include both pos+neg IP tracks; for negative JBP: use neg IP tracks only
+        Jet["Bprob"] = (
+            -np.log(jpc.calc_jet_proba(trk_Bproba_pos)) / 4.0
+            - np.log(jpc.calc_jet_proba(trk_proba_all)) / 4.0
+        )
+        Jet["BprobN"] = (
+            -np.log(jpc.calc_jet_proba(trk_Bproba_neg)) / 4.0
+            - np.log(jpc.calc_jet_proba(trk_proba_neg)) / 4.0
```

Comment:
`jpc.calc_jet_proba(trk_proba_neg)` appears twice in the code, I am wonder whether we can simplify this by `Jet["ProbaN"]*np.log(10)` to transfer to `- np.log(jpc.calc_jet_proba(trk_proba_neg)) / 4.0`

Reply:
Thanks for the clarification I misunderstood the differences between `track_proba_neg & track_jet_proba_neg`